### ISSUE TITLE
syntax: add mksh support

### ISSUE THREAD

**Issue by mvdan** [mvdan/sh#79]
Initially suggested by @D630 in #77.

Open questions needed to be answered before we decide whether or not to support it:

* Where is the formal specification of its syntax? The man page (https://www.mirbsd.org/htman/i386/man1/false.htm) seems to be the only reference.
* What language features does it have that we don't yet support via either POSIX Shell or Bash?
* Is it compatible with the POSIX Shell syntax? i.e., do they conflict?
* Is it compatible with the Bash syntax? i.e., do they conflict?

And of course, whether or not this is worth the effort. I myself haven't seen any `mksh` scripts in the wild, whereas `sh` and `bash` are both very common.

**Comment by D630** [comment]
cc @mirabilos

**Comment by mvdan** [comment]
Also, to clarify what I mean by compatibility with POSIX/Bash - Bash is technically not compatible with POSIX, but it's compatible enough that we can teach the parser Bash on top of POSIX without much effort. I'm wondering whether mksh could sit on top of POSIX as well, or even share some of the parser/lexer code with Bash.

**Comment by mvdan** [comment]
After a quick look at the man page:

* Shares quite a lot of language features with Bash like arrays, complex parameter expansions, `[[` and `function`
* Adds some language features like `${ foo;}` and `${|foo}` (not yet sure I understand their purpose)

All in all, seems like it's very similar to Bash and not incompatible syntax-wise.

**Comment by mirabilos** [comment]
`mksh` is a POSIX superset (except character classes in extglobs) with a lot of extensions, most shared with other Korn shells or GNU bash.

Note that mksh is able to convert code into a “canonical form” automatically, even disambiguating what your tool cannot (including reentry):

    tg@blau:~ $ foo() {
    > echo $((1+2))
    > echo $((date)|tr a-zA-Z A-Za-z)
    > }
    tg@blau:~ $ typeset -f foo
    foo() {
            \echo $((1+2))
            \echo $( ( \date ) | \tr a-zA-Z A-Za-z )
    }
    tg@blau:~ $ eval "$(typeset -f foo)"
    tg@blau:~ $ typeset -f foo
    foo() {
            \echo $((1+2))
            \echo $( ( \date ) | \tr a-zA-Z A-Za-z )
    }
    tg@blau:~ $ foo
    3
    tUE aPR 11 12:32:39 utc 2017

That’s CVS HEAD, older versions’ output differed as I fix bugs all the time.

You could use that (ugly but at least reentrant) output as the base for your own improvements on the formatting.

`${ foo;}` and `${|foo;}` (note the mandatory semicolon-or-newline) are inline function-like constructs that were created (the first by dgk, the second by me) to eliminate `fork()` overhead from `$(foo)`. The first has tempfile penalty (stdout is redirected to a file), the second requires the inner code to write to a special parameter `REPLY` instead of to stdout but is fastest. Parse them as if `foo` were a function body.

There’s no formal specification, none of the contemporary shells have one, and it’s living.

HTH & HAND

**Comment by mvdan** [comment]
Thanks for your time, @mirabilos.

> Note that mksh is able to convert code into a “canonical form” automatically, even disambiguating what your tool cannot (including reentry)

This parser used to deal with ambiguity via backtracking. But I chose to remove it because it was impossible to support properly while keeping the parser stream-based. See #30. Also see the caveats section in the README - these are tradeoffs made to keep the library saner.

> Parse them as if foo were a function body.

Thanks for the explanation. Seems like it shouldn't be hard, parser-wise.

> There’s no formal specification, none of the contemporary shells have one, and it’s living.

That's fine; was just asking in case there was one similar to the POSIX Shell spec.

Also, I just realised mksh is the current Android shell. That's a fairly big point in favor of its popularity and usefulness as a scripting language.

I think I'm convinced as far as supporting mksh in this package. I'll have to think about how to change the code to accomodate for another shell language variant.

One last question - I assume `mksh` is not a superset of Bash, correct?

**Comment by mirabilos** [comment]
OK, sounds sensible.

No, `mksh` is not a superset of GNU bash, although it does support some of its extensions, depending on the shell mode (the `&>foo` alias for `>foo 2>&1` is only supported in non-POSIX mode as it’s ambiguous when doing `ls&>echo foo`).

**Comment by mvdan** [comment]
> No, mksh is not a superset of GNU bash

Thanks for confirming. I was thinking that, but seeing all of the supported Bash features made me wonder.

> is only supported in non-POSIX mode as it’s ambiguous when doing `ls&>echo foo`

You made me jump for a second, but I forgot I already had a test case for this :)

I'll rename the issue. Not sure when I'll have time to work on this, but it should be soon™.

**Comment by mvdan** [comment]
Here's an interesting one - should the parser default to bash (what it does now), POSIX, or a superset of all shells (bash and mksh once this is done)?

In other words, would a mode that accepts both bash and mksh make sense at all? For the purposes of `shfmt`, perhaps - but even then you could argue that `shfmt` would lose its ability to tell if a program is valid syntax or not for its shell variant.

Leaning towards adding another parser mode, similar to `PosixConformant`, that switches from bash to mksh. Can't think of a proper name now, perhaps `MkshVariant` or `MkshMode`.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>Here's an interesting one - should the parser default to bash (what it
>does now), POSIX, or a superset of all shells (bash and mksh once this
>is done)?
>
>In other words, would a mode that accepts both bash and mksh make
>sense at all?

No, because mksh doesn’t equal mksh… there are two main variants (mksh
and lksh) which differ by which underlying integer type they use for
shell arithmetics, and then we have 'set -o sh' and 'set -o posix' and
invoking GNU bash as 'bash --posix' which also alter PARSING behaviour
(not just runtime).

>For the purposes of `shfmt`, perhaps - but even then you could argue
>that `shfmt` would lose its ability to tell if a program is valid
>syntax or not for its shell variant.

Good question. I’d probably let the user pass the exact mode as an
option, or try to divine from the shebang (#!/bin/{mk,ba}sh or
#!/usr/bin/env {mk,ba}sh enable the default extension modes, all
others use POSIX mode), and still offer other modes (like 'mksh-posix'
and 'bash-posix'; you’ll probably not have to consider lksh as
different).

>Leaning towards adding another parser mode, similar to
>`PosixConformant`, that switches from bash to mksh. Can't think of a
>proper name now, perhaps `MkshVariant` or `MkshMode`.

Please don’t uppercase anything in “mksh”, that makes me cringe.
If you must, write it out: “MirBSD Korn Shell”

Thanks,
//mirabilos
-- 
FWIW, I'm quite impressed with mksh interactively. I thought it was much
*much* more bare bones. But it turns out it beats the living hell out of
ksh93 in that respect. I'd even consider it for my daily use if I hadn't
wasted half my life on my zsh setup. :-) -- Frank Terbeck in #!/bin/mksh

**Comment by mvdan** [comment]
> No, because mksh doesn’t equal mksh… there are two main variants (mksh and lksh) which differ by which underlying integer type they use for shell arithmetics

Like you said, lksh would not need to be considered different than mksh in the syntax (lexer/parser/printer) package. This would only be a difference in the interpreter, which for now (still in progress) has no modes.

> I’d probably let the user pass the exact mode as an option, or try to divine from the shebang (#!/bin/{mk,ba}sh or#!/usr/bin/env {mk,ba}sh enable the default extension modes, all others use POSIX mode)

I thought about this a while ago (#48), but decided to keep it explicit via parameters since that keeps the tool simpler and I couldn't think of a scenario where you'd keep multiple languages in a bunch of shell scripts within a project.

This detection would also get wonkier with stdin, as I was thinking of including file extensions too. Perhaps they should be left out.

> and still offer other modes (like 'mksh-posix' and 'bash-posix')

How would these be useful for the syntax package? As far as I know, the language is either POSIX or Bash, not a mix.

> Please don’t uppercase anything in “mksh”, that makes me cringe.
If you must, write it out: “MirBSD Korn Shell”

I must; that's how names are exported in Go. And names are camelCase, so even if I use a "mode" prefix it would still be `ModeMksh` as `Modemksh` is really confusing.

If you strongly feel against `Mksh` in there, I can try to avoid it. How about `MirBSDKorn`, `MirBSDVariant` or `MirBSDKornVariant`? I know you added `Shell`, but I want to keep the constant short if possible.

These are integer masks, so the usage would be like `mode := ParseComments | MirBSDKorn`.

**Comment by mvdan** [comment]
Also, `MKSH` is of course an option but I think that would drive you up the wall even worse than `Mksh`.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>> No, because mksh doesn’t equal mksh… there are two main variants
>> (mksh and lksh) which differ by which underlying integer type they
>> use for shell arithmetics
>
>Like you said, lksh would not need to be considered different than

Yes.

>> I’d probably let the user pass the exact mode as an option, or try
>> to divine from the shebang (#!/bin/{mk,ba}sh or#!/usr/bin/env
>> {mk,ba}sh enable the default extension modes, all others use POSIX
>> mode)
>
>I thought about this a while ago (#48), but decided to keep it
>explicit via parameters since that keeps the tool simpler and I
>couldn't think of a scenario where you'd keep multiple languages in a
>bunch of shell scripts within a project.

OK, agreed.

>> and still offer other modes (like 'mksh-posix' and 'bash-posix')
>
>How would these be useful for the syntax package? As far as I know,
>the language is either POSIX or Bash, not a mix.

No.

If you use `bash --posix` or `mksh -o posix`, only _some_ extensions
(those that conflict with POSIX, plus the (re-enablable) csh-style brace
expansion (in mksh’s case)) are disabled, all others remain.

The `set -o sh` mode is more for legacy code.

Admittedly, the exact barrier between `-o posix`, `-o sh`, both and none
in mksh is not yet decided (although that’s on my definitive TODO for
the next release, which again is on my definitive TODO for the next
couple of days).

>I must; that's how names are exported in Go. And names are camelCase,

Urgs… ok…

>How about `MirBSDKorn`, `MirBSDVariant` or `MirBSDKornVariant`? I know

Looking good (MirBSDKorn, MirBSDPosix are the two you most likely need).

Thanks,
//mirabilos
-- 
„Cool, /usr/share/doc/mksh/examples/uhr.gz ist ja ein Grund,
mksh auf jedem System zu installieren.“
	-- XTaran auf der OpenRheinRuhr, ganz begeistert
(EN: “[…]uhr.gz is a reason to install mksh on every system.”)

**Comment by mvdan** [comment]
> Admittedly, the exact barrier between `-o posix`, `-o sh`, both and none in mksh is not yet decided (although that’s on my definitive TODO for the next release, which again is on my definitive TODO for the next couple of days).

I always assumed that Bash's `--posix` was meant to be "only accept POSIX". But then, for example, bash arrays are apparently OK:

```
 $ echo 'foo=(1 2); echo ${foo[@]}' | bash --posix
1 2
```

That's why I never considered a `bash-posix` mode. It's not clearly defined, I don't understand its purpose, and even if it was clearly defined I still wouldn't see its use for the syntax package. Either you want to enforce POSIX or Bash, but not a mix.

I feel similarly about `mksh-posix`, really. Why would anyone write and maintain scripts in these two in-between languages?

In any case, it's simpler for me to focus solely on `mksh` first.

> Looking good (MirBSDKorn, MirBSDPosix are the two you most likely need).

I'll go with `MirBSDKorn` then, thanks.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>That's why I never considered a `bash-posix` mode. It's not clearly
>defined

Well, there _is_ the source…

>I don't understand its purpose, and even if it was clearly defined I
>still wouldn't see its use for the syntax package. Either you want to
>enforce POSIX or Bash, but not a mix.
>
>I feel similarly about `mksh-posix`, really. Why would anyone write
>and maintain scripts in these two in-between languages?

Let me ask the same question: “Why would anyone write anything in
Issue 9 (Google-Go)?”

But it’s probably okay if you concentrate on the main families first.

**Comment by mvdan** [comment]
> Well, there _is_ the source…

It's possible for me to do it for sure, I was more saying that it's of little use as a standalone language because developers can't easily know what it even has and doesn't have.

> Let me ask the same question: “Why would anyone write anything in Issue 9 (Google-Go)?”

What I'm getting at is that you either write in mksh and then require mksh to run, or you write in POSIX and then any POSIX-compatible shell can execute it. Writing in a language that is in between means that mksh will still be needed to run. Where is the benefit to writing a script in that language?

I'm not saying it's pointless. I'm saying that, from the point of view of this package and the tools I'm developing, I don't see why people wouldn't just use the `mksh` mode for `mksh-posix` programs. It would work all the same.

I'm not sure I understand your counter-question. I'm talking about languages that are in between two other languages, like in between mksh and POSIX Shell. Go is not in between of anything :)

**Comment by mirabilos** [comment]
You can easily merge POSIX stuff (e.g. shell functions) while retaining the ability to use the advanced features.

**Comment by mvdan** [comment]
Wait, so the language `mksh-posix` is a superset of `mksh`? I thought the opposite was true.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>Wait, so the language `mksh-posix` is a superset of `mksh`? I thought
>the opposite was true.

Neither is true.

They are very close, though.

bye,
//mirabilos
-- 
<igli> exceptions: a truly awful implementation of quite a nice idea.
<igli> just about the worst way you could do something like that, afaic.
<igli> it's like anti-design.  <mirabilos> that too… may I quote you on that?
<igli> sure, tho i doubt anyone will listen ;)

**Comment by mvdan** [comment]
I'm a bit confused :) For now I'll just work on the regular mksh language. Thanks for all your help!

**Comment by mvdan** [comment]
@mirabilos I'm starting work on this now. Do you have any test suite that I could use to get `mksh` shell files that should and should not error?

**Comment by mvdan** [comment]
I've ran my POSIX test suite on `mksh` to get a general idea. Some things to note:

* It errors if a heredoc isn't closed. I love you for this, no other shell seems to do this for some god damned reason. I might make my parser do the same.
* It seems to accept some invalid func declarations like `""() { x; }` (invalid name) and `foo()` (missing body)
* It seems to accept arithmetic assignments to non-idents, like `$(((a)=b))`. But it does accept `a=b` and reject `1=2`, so I guess it reduces `(a)=b` to `a=b` before checking if the value on the left is an ident. Is that correct? All other shells I've tried reject `(a)=b`.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>@mirabilos I'm starting work on this now. Do you have any test suite
>@that I could use to get `mksh` shell files that should and should not
>@error?

Well, the mksh testsuite (`check.t`) contains tons of examples.

Some more real-world scripts are in:
https://github.com/mirabilos/shellsnippets

“mksh -n file” is a quick (but not thorough, especially not wrt. eval)
syntax check on a file (basically, it parses the script but does not
execute it).

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>* It seems to accept some invalid func declarations like
> `""() { x; }` (invalid name)

	tg@blau:~ $ ""() { x; }
	tg@blau:~ $ typeset -f
	() {
		\x
	}

Indeed, this seems to be a bug, thanks for spotting.

Hey, no, wait!

	tg@blau:~ $ ""() { x; }
	tg@blau:~ $ ""
	/bin/mksh: x: not found

This looks deliberate. Heh. Weird. I guess quoted things go.

“invalid name” is disputable… POSIX allows extensions (we have
more chars allowed in alias names than POSIX does, for example).

> and `foo()` (missing body)

No, it reads a body:

	tg@blau:~ $ foo()
	> ;

In some old standard/shell, just “funcname() command” was okay.
If stdin closes (like in “mksh -c 'foo()'”) it’s an empty command.

>* It seems to accept arithmetic assignments to non-idents, like
> `$(((a)=b))`. But it does accept `a=b` and reject `1=2`, so I
> guess it reduces `(a)=b` to `a=b` before checking if the value on the
> left is an ident. Is that correct?

As far as I can tell, yes. There is nothing like C’s lvalues in
shell, and the variable can be assigned to.

**Comment by mvdan** [comment]
> Well, the mksh testsuite (`check.t`) contains tons of examples.

Thanks - will check those out.

> “mksh -n file” is a quick (but not thorough, especially not wrt. eval) syntax check on a file (basically, it parses the script but does not execute it).

I do use that on `bash` and `dash`, but unfortunately many shells don't fully parse the input unless they actually run it, like:

```
 $ echo '${foo:1 2}' | mksh
mksh: <stdin>[1]:  1 2: unexpected '2'
 $ echo '${foo:1 2}' | mksh -n
 $ echo 'exit; ${foo:1 2}' | mksh
```

I assume that's what the POSIX spec says, but I made my parser/tokenizer completely static. So I do use `-n` on inputs that should not error, but I don't use it on inputs that should error - to make sure the shells actually try to run the code.

> This looks deliberate. Heh. Weird. I guess quoted things go.
> “invalid name” is disputable… POSIX allows extensions (we have more chars allowed in alias names than POSIX does, for example).

(To clarify - I was bringing these up because I was curious and because they might be bugs in `mksh`. I'm not implying you should fix them.)

I was aware of extra characters like how Bash allows `-` and `+`. But I always assumed it was within a literal, i.e. not involving characters like `"`, `$` or `(`. In any case, a func with the name `""` doesn't make sense.

> In some old standard/shell, just “funcname() command” was okay. If stdin closes (like in “mksh -c 'foo()'”) it’s an empty command.

This parser does allow `foo() bar` as well as `foo() (bar)` and `foo() { bar; }`, but it doesn't accept `foo()`.

I've had a look at the latest POSIX spec, and it seems like only a brace group, a subshell or a clause (if/for/etc) are valid function bodies. It doesn't seem to say that no body is acceptable.

But as with the case above, I don't think I want to accept a function with no body because it makes little sense.

> There is nothing like C’s lvalues in shell, and the variable can be assigned to.

Grah! I think I'll leave this parser without the extra magic then. After all, most shells barf on `(a)=b` and I want this parser to be strict, in the sense that if it accepts a program every popular shell should too. And once more, `(a)=b` makes little sense when you can just write `a=b`.

Also, I've been looking at the heredoc stuff and I'm fairly sure the POSIX spec doesn't say anything about unclosed here-documents. I wonder why `bash`, `dash` and most shells are okay with this.

**Comment by mvdan** [comment]
@mirabilos here's one that looks more like an actual mksh bug - `[]` is accepted as an array index:

```
 $ bash <<<'echo ${foo[]}'
bash: line 1: ${foo[]}: bad substitution
 $ mksh <<<'echo ${foo[]}'
 $ bash <<<'echo $((foo[]))'
bash: line 1: foo[]: bad array subscript
bash: line 1: foo[]: bad array subscript
0
 $ mksh <<<'echo $((foo[]))'
0
```

I did not find anything in `man mksh` that could explain why it would be a valid program. Note that whether or not `foo` is declared (as an array) does not change anything.

**Comment by mvdan** [comment]
Here's another one, but I'm sure this is on purpose - `let a=(b)` is not allowed. I assume this is because you want the expression quoted in mksh. I'm bringing this up because Bash does support it, as does my parser. In fact, in my parser I forbid quotes in expressions to make sure the static parser works through all the expression, not letting anything to be done at run-time.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>@mirabilos here's one that looks more like an actual mksh bug - `[]`
>@is accepted as an array index:

Again, surprising me, but this is something I can explain:

In `${foo[bar]}`, `bar` is evaluated as arithmetic expression.
Here, `bar` is the empty string; however, that’s a valid arithmetic
expression evaluating to 0.

    tg@blau:~ $ echo $(())
    0

>Here's another one, but I'm sure this is on purpose - `let a=(b)` is

Yes, the opening parenthesis is a reserved word and may thus not
be unquoted on the command line, just like this is an error:

    tg@blau:~ $ echo foo(bar)baz
    /bin/mksh: syntax error: unexpected '('

>not allowed. I assume this is because you want the expression quoted
>in mksh. I'm bringing this up because Bash does support it, as does my

I think GNU bash _might_ be using this as array reference. I don’t know
it well, but AT&T ksh93 at least requires

    typeset -a foo=(bar baz)

where GNU bash uses (and recent mksh supports)

    foo=(bar baz)

and all Korn shells (including ksh88 and pdksh) use

    set -A foo -- bar baz

>parser. In fact, in my parser I forbid quotes in expressions to make
>sure the static parser works through all the expression, not letting
>anything to be done at run-time.

Unfortunately, expressions must be evaluated at runtime if they can’t
be analysed statically.

    (( foo = bar ))

is changed into

    \\builtin let ' foo = bar '

by mksh internally, so these two are equivalent. However, POSIX says
that this must work:

    tg@blau:~ $ x='2+3'
    tg@blau:~ $ echo $(($x*4))
    14

So you can’t even build a parse tree for expressions that have
substitutions in them (I was considering storing expressions in
a parsed form that can be evaluated faster internally; this is
_probably_ still possible, but only when there’s no $ in them).

And yes, those variables can even contain parenthesēs…

**Comment by mvdan** [comment]
> however, that’s a valid arithmetic expression evaluating to 0.

You're right; if I am being strict and not allowing `[]`, I shouldn't allow `$(())` either. I assume this is important if you want to allow substitutions in expressions, to end up with `0` rather than an error.

> the opening parenthesis is a reserved word and may thus not be unquoted on the command line

You're correct in that Bash reads it as an array. I've basically taught my parser to accept `foo=(bar)` if at the start of a statement, or if it's in a `declare`/`typeset`/etc. This is somewhat ugly, but it's Bash.

Regarding your point on POSIX and substitutions inside expressions; I agree and I see why you're doing it like that. But my parser being fully static, I just can't leave anything for the run-time. So you could say it breaks with POSIX, Bash and other languages. But I already do that by not backtracking and many others :)

I'd like to think that this forces saner shell programs. I'm afraid, though, that if `let a=(b + c)` must be quoted in `mksh`, our two programs will be mutually exclusive. Not sure what I can do about that.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>> “mksh -n file” is a quick (but not thorough, especially not wrt.
>> eval) syntax check on a file (basically, it parses the script but
>> does not execute it).
>
>I do use that on `bash` and `dash`, but unfortunately many shells
>don't fully parse the input unless they actually run it, like:
>
>```
> $ echo '${foo:1 2}' | mksh
>mksh: <stdin>[1]:  1 2: unexpected '2'
> $ echo '${foo:1 2}' | mksh -n
> $ echo 'exit; ${foo:1 2}' | mksh
>```
>
>I assume that's what the POSIX spec says

Exactly. I’ve even got a usecase for it: a shell tester.
“What shell am I running under?”

It’s capable of testing most Bourne/POSIX shells, C shells
and even Wish (the Tcl/Tk shell). Just invoke it like
`mksh getshver` or `csh getshver` or `wish getshver`.

http://www.mirbsd.org/cvs.cgi/contrib/code/Snippets/getshver?rev=HEAD

Each shell only parses what they need, for example, Tcl resolves
backslash+newline before comments, so any comment ending with a
backslash hides the next line from it, whereas shell runs it.
The C shell has goto (yay!) so it’s at the end, which it just
skips without parsing, and the “exit 0” above terminates all
Bourne/POSIX shells. Only the ksh93 `${.sh.version}` is tested
with eval; usually I just do something like this to skip, for
example in `~/.profile`:

    # aliases, functions, etc. that are POSIX compatible
    test -n "$KSH_VERSION" || return 0
    # aliases, functions, etc. that require mksh extensions

>but I made my parser/tokenizer completely static. So I do use `-n` on

I bet “getshver” would cause your code to explode ;) but in the
scope you have, making it completely static is the only thing to
do to keep one’s sanity.

As with the arithmetic thing: it’s impossible to say whether…

    x=$(somecommand)
    echo $(($x + 1))

… is valid or not without actually running the code, period.
There are limitations to static testing, and the only thing
you can do is to document them (and that’s fully okay).

>(To clarify - I was bringing these up because I was curious and
>because they might be bugs in `mksh`.

Yes, indeed… and here, my own (well, inherited, but I worked
on it for about 15 years, give or take, now) shell surprised
myself. I’m thankful for you bringing them up.

>In any case, a func with the name `""` doesn't make sense.

The name is actually the empty string… with just `''` you
can also invoke it.

Handy, right? ;-) I can imagine one use case.

>> In some old standard/shell, just “funcname() command” was okay. If
>> stdin closes (like in “mksh -c 'foo()'”) it’s an empty command.
>
>This parser does allow `foo() bar` as well as `foo() (bar)` and `foo()
>{ bar; }`, but it doesn't accept `foo()`.

That’s likely correct.

>But as with the case above, I don't think I want to accept a function
>with no body because it makes little sense.

Yes.

>Grah! I think I'll leave this parser without the extra magic then.
>After all, most shells barf on `(a)=b` and I want this parser to be
>strict

Agreed.

>Also, I've been looking at the heredoc stuff and I'm fairly sure the
>POSIX spec doesn't say anything about unclosed here-documents. I
>wonder why `bash`, `dash` and most shells are okay with this.

Good question… I have no idea, but Chet Ramey (the GNU bash developer)
is pretty reachable. (Feel free to Cc me.)

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>I'm afraid, though, that if `let a=(b + c)` must be quoted in `mksh`,

It must be quoted in POSIX shell.

You can, however, write: `let a=b+c`

You can also write multiple evaluations in one `let` statement:
`let a=b+c d=e+f`

I’d say what you can do is either to ignore a quoted argument
to let, or to handle the content of the quotes as if unquoted.
I think the latter should work pretty well (besides, I think
that `let arrayname=(value)` is a semantic mistake in GNU bash).
`let` is not a special command like `typeset` or `declare`.

Most mksh users will write `(( a = (b + c) ))` however, and,
if at all, use `let` only for stuff like `let ++x`, so we
should have no problems even if you decide to just leave
quoted `let` arguments alone instead of peeking into the quotes.

**Comment by mvdan** [comment]
> I’ve even got a usecase for it: a shell tester.

That's insane and beautiful. Why not just pattern match `$KSH_VERSION`, `$BASH_VERSION` and such, though?

> As with the arithmetic thing: it’s impossible to say whether

Yup. I do allow `${}` for some edge cases, but the parser just assumes that the result will be a single expression like a name or a number. If `${foo}` results in `1 + 2`, it will break. I don't want to get into the territory of the interpreter having to re-run the parser on each expression. If someone really needs that, they should just `eval`.

> I can imagine one use case.

For an empty-named func? I actually don't see it :)

> I have no idea, but Chet Ramey (the GNU bash developer) is pretty reachable.

Actually, Bash does detect it and warn, but it lets it through. This is very likely for backwards compat reasons, as the shells it tends to replace in many Linux distros like Dash don't even do a warning (both have exit code 0):

```
 $ bash <<<'<<EOF'
bash: line 1: warning: here-document at line 1 delimited by end-of-file (wanted `EOF')
 $ dash <<<'<<EOF'
```

> a semantic mistake in GNU bash

Welcome to my world of pain! Most of these things aren't properly documented. For instance, just yesterday I playing with Bash to reverse-engineer (don't want to read its code) to figure out whether `coproc name foo | bar` means a coprocess with the program `foo | bar`, or just `foo`. The manual just says "command", which is as useful as it gets.

In general I agree with you that `let` is weird. It's supposedly a builtin, but as you say it acts as a keyword that modifies how the parser/tokenizer work. I've basically come to terms with that, and allow the use of parentheses there to be closer to bash. After all, `let a=(b-c)+c` is not that bad if you overlook how `let` is supposed to be a builtin.

You're right that in the case of my parser versus mksh, one can just rewrite the expression to avoid parentheses. I'll take that as enough reason to think I don't have to fix this conflict.

As to parsing the content of quotes... That seems like a rabbit hole of complexity in the AST, especially when it comes to printing it back and with teaching the parser and lexer to treat quotes differently in that situation. I'll overlook this for now. I might come back to this idea if anyone nags me about it and raises good points.

Thanks again for all your help! This is being really helpful.

If you look at the commits above, I'm pretty much done with the `mksh` features that it shares with Bash. As to the features of its own, I've only done `|&` so far. Need to do the `${ foo;}` and `${|foo}` variations. After that, it will just be a matter of running my parser on your scripts and play whack-a-mole with the inconsistencies.

**Comment by mirabilos** [comment]
Hi,

>> I’ve even got a usecase for it: a shell tester.
>
>That's insane and beautiful. Why not just pattern match
>`$KSH_VERSION`, `$BASH_VERSION` and such, though?

that isn’t sufficient because those are a recent invention,
and the shells you’re usually interested in distinguishing
because of, ahem, an “interesting” interpretation of the
standard (or even being pre-POSIX) don’t have them.

>> As with the arithmetic thing: it’s impossible to say whether
>
>Yup. I do allow `${}` for some edge cases, but the parser just assumes
>that the result will be a single expression like a name or a number.
>If `${foo}` results in `1 + 2`, it will break. I don't want to get

Sure.

>into the territory of the interpreter having to re-run the parser on
>each expression. If someone really needs that, they should just
>`eval`.

But eval’s the wrong tool for that…

>> I can imagine one use case.
>
>For an empty-named func? I actually don't see it :)

Running input as functions, to recognise empty input,
or something. Nothing really fancy.

>`coproc name foo | bar` means a coprocess with the program `foo |
>bar`, or just `foo`. The manual just says "command", which is as
>useful as it gets.

Ouch, yes. I’ve not understood how GNU bash does coprocesses either.

>In general I agree with you that `let` is weird. It's supposedly a
>builtin, but as you say it acts as a keyword that modifies how the
>parser/tokenizer work.

No, neither in bash (it always parses x=(y) as array assignment)
nor in mksh (you do need to quote).

>allow the use of parentheses there to be closer to bash. After all,
>`let a=(b-c)+c` is not that bad if you overlook how `let` is supposed
>to be a builtin.

But it’s a mistake for a script to write that without quoting.
They really want a=$(((b-c)+c)) instead of a[0]='b-c', and I’ve
no idea where the +c would end up.

So I’d throw an error if there are unquoted parenthesēs, just
like mksh does, because even if bash accepts the input, it
does not do what it’s meant to.

>Thanks again for all your help! This is being really helpful.

Thanks you too.

>If you look at the commits above, I'm pretty much done with the `mksh`
>features that it shares with Bash. As to the features of its own, I've
>only done `|&` so far. Need to do the `${ foo;}` and `${|foo}`
>variations. After that, it will just be a matter of running my parser
>on your scripts and play whack-a-mole with the inconsistencies.

Ah, I see. Have fun then. ${ˤfoo;} for ˤ in (space, tab, newline, bar)
can be parsed just as a function body. The trailing semicolon before
the closening curly brace can also be a newline, mind you, it’s only
necessary due to the function-body-like syntax.

**Comment by mvdan** [comment]
> But eval’s the wrong tool for that

How so? In my eyes, `$(($containsexpr))` is forcing the parser to run at run-time, which would otherwise only happen with `eval`.

> No, neither in bash (it always parses x=(y) as array assignment)

I think you're wrong:

```
 $ ((x=(1 + 2))); echo $x
3
 $ let x=(1 + 2); echo $x
3
 $ x=(1 + 2); echo $x # first element
1
```

> But it’s a mistake for a script to write that without quoting.

In POSIX Shell sure - but not in Bash, in practice. In the example I showed above, Bash does behave like it's supposed to. Unless you think it should parse `let x=(y)` as an array, which is wrong because assignments must go before any argument:

```
 $ foo x=(y)
bash: syntax error near unexpected token `('
```

**Comment by mvdan** [comment]
Here's a formatting question. I see you write `${ foo;}` and `${|foo;}`, not `${ foo; }`, like one would when using a single-line block like `foo() { bar; }`. Is there a reason?

I can make the printer/formatter do either `;}` or `; }` for this specific case when it's in a single line, I'm not too bothered. Just wondering about it since `; }` would be more consistent with elsewhere in the format this tool currently enforces.

**Comment by mvdan** [comment]
I've gone for no spacing for now, following `man mksh`, though the opposite would seem like the better option for consistency.

I've closed this issue but I'm sure I've missed a few things and there are bugs and edge cases. If you find any, please file separate issues.

**Comment by mvdan** [comment]
I used the shellsnippets repo and found two very common missing features:

https://github.com/mvdan/sh/issues/111
https://github.com/mvdan/sh/issues/112

So not particularly useful just yet.

**Comment by mvdan** [comment]
FYI, all issues that I could understand are now fixed. `shellsnippets/mksh` is almost entirely parsed correctly, minus a few known limitations and `c == 1#%` which I haven't found the meaning for.

I'm deeming this good for 2.0, so the time to test is now - before 2.0 is released and I have to be careful with backwards compat :)

**Comment by mvdan** [comment]
@mirabilos abusing this thread a bit for a small issue I encountered:

```
$ bash -c "foo=(a b) env | grep '^foo='"
foo=(a b)
$ mksh -c "foo=(a b) env | grep '^foo='"
mksh: syntax error: unexpected '|'
```

I know that passing non-string variables to a command is a bit silly, but at least the error is very confusing. It's also worth pointing out that `mksh -c "foo=(a b) env"` prints nothing, almost as if it were interpreting `env` as a variable declaration.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>@mirabilos abusing this thread a bit for a small issue I encountered:
>
>```
>$ bash -c "foo=(a b) env | grep '^foo='"
>foo=(a b)
>$ mksh -c "foo=(a b) env | grep '^foo='"
>mksh: syntax error: unexpected '|'
>```
>
>I know that passing non-string variables to a command is a bit silly,
>but at least the error is very confusing. It's also worth pointing out
>that `mksh -c "foo=(a b) env"` prints nothing, almost as if it were
>interpreting `env` as a variable declaration.

Hm. Well, “foo=(a b)” is translated to “set -A foo -- a b”.
Handling of anything else on the command line is… suboptimal.

GNU bash’s output is also wrong, though. It’s supposed to
be $'foo[0]=a\nfoo[1]=b' or a syntax error, depending.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>It's also worth pointing out that `mksh -c "foo=(a b) env"` prints

Thanks, you spotted a rather deep redesign bug, which to debug
cost me several hours today alone to get the debugging invocations
right, today.

>nothing, almost as if it were interpreting `env` as a variable
>declaration.

It just ignored it instead of requiring a command end (semicolon,
newline or end of input). Basically, any LWORD was ignored :(

**Comment by mvdan** [comment]
Great to hear. I also made `foo=(a b) prog` a syntax error, basically saying that arrays cannot be inline variables.

I agree that what bash does is weird. In practice, what it does is the same as if the program were `foo='(a b)' prog`. Which doesn't make sense in any situation, in my opinion. I guess it's a byproduct of the shell "stringifying" all variables before passing them onto a forked process.

**Comment by mvdan** [comment]
By the way, since you might be interested in other potential bugs, here is one:

```
$ ! !
mksh: syntax error: unexpected 'newline'
```

I know this is a silly program, but this should execute like `! '!'`. That is, the `!` program with a negated exit code. Bash and my parser get this right, and `dash` and `mksh` get it wrong. That is because `!` is simply a reserved word, and not a special token in any way. For example, `mksh` runs `echo !` just fine.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>Great to hear. I also made `foo=(a b) prog` a syntax error, basically
>saying that arrays cannot be inline variables.

Indeed. That might eventually appear, but not now.

>I agree that what bash does is weird. In practice, what it does is the
>same as if the program were `foo='(a b)' prog`. Which doesn't make
>sense in any situation, in my opinion. I guess it's a byproduct of the

Exactly.

>By the way, since you might be interested in other potential bugs, here is one:
>
>```
>$ ! !
>mksh: syntax error: unexpected 'newline'
>```

No, that is correct.

>I know this is a silly program, but this should execute like `! '!'`.
>That is, the `!` program with a negated exit code. Bash and my parser

No.

tg@blau:~ $ ! ! true; echo = $?
= 0
tg@blau:~ $ ! ! false; echo = $?
= 1
tg@blau:~ $ ! ! (exit 12); echo = $?
= 1

You can have multiple negations.

>For example, `mksh` runs `echo !` just fine.

I’ve had problems with some bash versions doing
`if ! test foo = bar; then …` in which it interpreted
the `!` as history recall; mksh, in general, is much
more accepting of the exclamation sign.

[POSuX](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_04) agrees:

> This recognition shall only occur when none of the characters is quoted and when the word is used as:
>  * The first word of a command
>  * The first word following one of the reserved words other than `case`, `for`, or `in`

So: first word is the first `!`, which is “a reserved word other than…”,
which is followed by another `!`, which then, in my reading of it, even
*must* be taken as reserved word (this lets me think GNU `bash` and
`dash` have a bug here).

Heck, I think they do (or, at least, your interpretation is wrong):

    debian$ \which -a !
    debian$ bash -c '! !'; echo = $?
    = 0

This must give 127, command not found, when running an `!` utility.

Interestingly enough, the shell grammar in POSIX forbids this…

    pipeline         :      pipe_sequence
                     | Bang pipe_sequence
                     ;

… but it’s non-normative, informative only, and it’s well-known
(and, I think, even documented by POSIX) that it’s buggy and
non-compliant.

**Comment by mvdan** [comment]
Thanks for quoting the spec - I read all this a bit too fast when implementing it, I think.

And my assumption is wrong. Bash doesn't parse `! !` as negation of the `!` program. It does the same thing that `mksh` does, but it allows for an empty command after it.

```
$ ! ; echo $?
1
$ ! ! ; echo $?
0
$ ! ! ! ; echo $?
1
```

So the distinction between bash and mksh is simply whether or not an empty command (an empty pipe_sequence, to be exact) should be allowed. I have found these "empty list" rules to be very chaotic in the past, so I haven't paid much attention to them.

One example - bash, the very same shell that allows an empty pipe sequence there, forbids an empty list of commands (compound-list) here:

```
$ if; then echo foo; fi
bash: syntax error near unexpected token `;'
```

Assuming your reading of the spec is correct, then my AST and parser are wrong. But it leaves me to wonder if multiple negations are ever useful. I guess two could be used to convert all non-zero exit codes to 1, but that's a rather obscure feature to use.

In any case, parsing `! !` like `! '!'` is definitely wrong - I should at least change it to be an error. Unfortunately, I made the leading `!` a `Negated bool` in a flattened command struct, so I can't go back and add multiple levels of negation without breaking backwards compatibility in the AST.

**Comment by mirabilos** [comment]
Daniel Martí dixit:

>So the distinction between bash and mksh is simply whether or not an
>empty command (an empty pipe_sequence, to be exact) should be allowed.
>I have found these "empty list" rules to be very chaotic in the past,
>so I haven't paid much attention to them.

Yeah, I’ve not bothered much with them either, as mksh inherited
from pdksh, which was good enough.

I’d assume the empty command to be a GNU bash extension.

>But it leaves me to wonder if multiple negations are ever useful. I

They are, I’ve used them in very dynamic code.

>guess two could be used to convert all non-zero exit codes to 1, but

That too (often seen in C).

bye,
//mirabilos
-- 
“It is inappropriate to require that a time represented as
 seconds since the Epoch precisely represent the number of
 seconds between the referenced time and the Epoch.”
	-- IEEE Std 1003.1b-1993 (POSIX) Section B.2.2.2
